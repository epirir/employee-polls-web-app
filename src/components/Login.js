import { connect } from "react-redux";
import { Navigate } from "react-router-dom";
import { useState } from "react";
import { handleLogin } from "../actions/authedUser";

const Login = ({ dispatch, loggedIn }) => {
  const [username, setUsername] = useState("sarahedo");
  const [password, setPassword] = useState("password123");

  if (loggedIn) {
    const urlSearchParams = new URLSearchParams(window.location.search);
    const redirectToUrl = urlSearchParams.get("redirectTo");
    return <Navigate to={redirectToUrl ? redirectToUrl : "/"} />;
  }

  const onChangeUsername = (e) => {
    const value = e.target.value;
    setUsername(value);
  };

  const onChangePassword = (e) => {
    const value = e.target.value;
    setPassword(value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    dispatch(handleLogin(username, password));
    setUsername("");
    setPassword("");
  };

  return (
    <div className="bg-white shadow p-12 mt-32 max-w-sm mx-auto rounded">
      <h1
        className="text-3xl font-bold mt-9 text-center"
        data-testid="login-heading"
      >
        Login
      </h1>
      <form onSubmit={handleSubmit}>
        <div>
          <label
            htmlFor="username"
            className="block text-sm font-medium text-slate-700"
          >
            Username
          </label>
          <div className="mt-1">
            <input
              value={username}
              onChange={onChangeUsername}
              type="text"
              name="username"
              id="username"
              data-testid="username"
              placeholder="Type username..."
              className="mb-4 p-2 appearance-none block w-full bg-gray-200 placeholder-gray-500 rounded border focus:border-teal-500 placeholder-italic"
            />
          </div>
        </div>
        <div className="mt-6">
          <label
            htmlFor="password"
            className="block text-sm font-medium text-slate-700"
          >
            Password
          </label>
          <div className="mt-1">
            <input
              value={password}
              onChange={onChangePassword}
              type="password"
              name="password"
              id="password"
              data-testid="password"
              placeholder="Type password..."
              className="mb-4 p-2 appearance-none block w-full bg-gray-200 placeholder-gray-500 rounded border focus:border-teal-500 placeholder-italic"
            />
          </div>
        </div>
        <div className="mt-6 text-center">
          <button
            type="submit"
            data-testid="submit"
            className="w-full p-3 mt-4 bg-indigo-600 text-white rounded shadow"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  );
};

const mapStateToProps = ({ authedUser }) => ({
  loggedIn: Boolean(authedUser),
});

export default connect(mapStateToProps)(Login);
