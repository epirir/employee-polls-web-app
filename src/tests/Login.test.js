import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import { store } from "../store";
import { BrowserRouter } from "react-router-dom";
import React from "react";
import Login from "../components/Login";
import { handleInitialData } from "../actions/shared";

describe("Login", () => {
  it("should render the component", () => {
    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );
    expect(view).toBeDefined();
    expect(view).toMatchSnapshot();
  });

  it("should clear input elements after clicking submit button", async () => {
    await store.dispatch(handleInitialData());

    const view = render(
      <Provider store={store}>
        <BrowserRouter>
          <Login />
        </BrowserRouter>
      </Provider>
    );

    const testIds = ["login-heading", "username", "password", "submit"];

    testIds.forEach((testId) => {
      expect(view.getByTestId(testId)).toBeInTheDocument();
    });

    const username = view.getByTestId("username");
    const password = view.getByTestId("password");
    const submit = view.getByTestId("submit");
    const loginHeading = view.getByTestId("login-heading");

    fireEvent.change(username, { target: { value: "sarahedo" } });
    fireEvent.change(password, {
      target: { value: "wrongpassword" },
    });
    expect(username.value).toBe("sarahedo");
    expect(password.value).toBe("wrongpassword");
    fireEvent.click(submit); // User stays on page
    expect(loginHeading).toBeInTheDocument();
    expect(username.value).toBe("");
    expect(password.value).toBe("");
  });
});
