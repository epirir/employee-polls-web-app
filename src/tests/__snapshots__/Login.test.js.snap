// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="bg-white shadow p-12 mt-32 max-w-sm mx-auto rounded"
      >
        <h1
          class="text-3xl font-bold mt-9 text-center"
          data-testid="login-heading"
        >
          Login
        </h1>
        <form>
          <div>
            <label
              class="block text-sm font-medium text-slate-700"
              for="username"
            >
              Username
            </label>
            <div
              class="mt-1"
            >
              <input
                class="mb-4 p-2 appearance-none block w-full bg-gray-200 placeholder-gray-500 rounded border focus:border-teal-500 placeholder-italic"
                data-testid="username"
                id="username"
                name="username"
                placeholder="Type username..."
                type="text"
                value="sarahedo"
              />
            </div>
          </div>
          <div
            class="mt-6"
          >
            <label
              class="block text-sm font-medium text-slate-700"
              for="password"
            >
              Password
            </label>
            <div
              class="mt-1"
            >
              <input
                class="mb-4 p-2 appearance-none block w-full bg-gray-200 placeholder-gray-500 rounded border focus:border-teal-500 placeholder-italic"
                data-testid="password"
                id="password"
                name="password"
                placeholder="Type password..."
                type="password"
                value="password123"
              />
            </div>
          </div>
          <div
            class="mt-6 text-center"
          >
            <button
              class="w-full p-3 mt-4 bg-indigo-600 text-white rounded shadow"
              data-testid="submit"
              type="submit"
            >
              Login
            </button>
          </div>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="bg-white shadow p-12 mt-32 max-w-sm mx-auto rounded"
    >
      <h1
        class="text-3xl font-bold mt-9 text-center"
        data-testid="login-heading"
      >
        Login
      </h1>
      <form>
        <div>
          <label
            class="block text-sm font-medium text-slate-700"
            for="username"
          >
            Username
          </label>
          <div
            class="mt-1"
          >
            <input
              class="mb-4 p-2 appearance-none block w-full bg-gray-200 placeholder-gray-500 rounded border focus:border-teal-500 placeholder-italic"
              data-testid="username"
              id="username"
              name="username"
              placeholder="Type username..."
              type="text"
              value="sarahedo"
            />
          </div>
        </div>
        <div
          class="mt-6"
        >
          <label
            class="block text-sm font-medium text-slate-700"
            for="password"
          >
            Password
          </label>
          <div
            class="mt-1"
          >
            <input
              class="mb-4 p-2 appearance-none block w-full bg-gray-200 placeholder-gray-500 rounded border focus:border-teal-500 placeholder-italic"
              data-testid="password"
              id="password"
              name="password"
              placeholder="Type password..."
              type="password"
              value="password123"
            />
          </div>
        </div>
        <div
          class="mt-6 text-center"
        >
          <button
            class="w-full p-3 mt-4 bg-indigo-600 text-white rounded shadow"
            data-testid="submit"
            type="submit"
          >
            Login
          </button>
        </div>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
